lib/markdownlint.js

```diff
@@ -895,6 +895,39 @@ function parseConfiguration(name, content, parsers) {
   };
 }
 
+/**
+ * Resolve referenced "extends" path ina configuration file
+ * using path.resolve() and require.resolve() as a fallback.
+ *
+ * @param {string} configFile Configuration file name.
+ * @param {*} referenceFile Referenced path to resolve.
+ * @returns {string}
+ */
+function resolveConfigExtends(configFile, referenceFile) {
+  const resolvedExtendsFile = path.resolve(
+    path.dirname(configFile),
+    referenceFile
+  );
+
+  try {
+    if (fs.statSync(resolvedExtendsFile).isFile()) {
+      return resolvedExtendsFile;
+    }
+  // eslint-disable-next-line unicorn/prefer-optional-catch-binding
+  } catch (error) {
+    // If fs.statSync has thrown, trying require.resolve
+  }
+
+  try {
+    return require.resolve(path.dirname(configFile), referenceFile);
+  // eslint-disable-next-line unicorn/prefer-optional-catch-binding
+  } catch (error) {
```

Comment:
Are these catches to be extra safe, or would both methods throw in normal use? I prefer to let true exceptions bubble out.

Reply:
The key principle I was trying to follow is backwards-compatibility, including the kinds of error messages people get for the unresolved paths. This is what resulted this logic.

If the original `path.resolve` could not throw, `resolveConfigExtends` should not either. It should first try to resolve the reference using `path.resolve` as before _and only if_ the resulting path does not point to an existing _file_ (not folder), try `require.resolve`. By default, `require.resolve` throws, so if we re-throw, the resulting error message ends up being different from previous versions. Thus, we’d have to qualify this PR as a breaking change, which I don’t think is necessary.

That said, I’m open to suggestions 🙂

Reply:
I'm thinking of this as a feature and want it to go out as a semver-breaking change. So I'm not worried about the kind of exception changing. Everything that used to work should continue to behave the same - new scenarios can start working - but things that didn't work can not work differently now. :)

Reply:
IC. How would you implement resolution? There are several ways, each with its own trade-offs. So I'm curious to know your vision, including the priority of the resolutions and the expected feedback to users.

Reply:
What you have here seems good! I'd prefer to get rid of the ESLint suppression's, I think the `error` parameter is unnecessary?

Reply:
I tried a few things here, but could not get rid of an ESLint error. Having `} catch {` instead of `} catch (error) {` produces:

```
Parsing error: Unexpected token { eslint
```

Adding it back gives:

```
Remove unused catch binding `error`. eslint(unicorn/prefer-optional-catch-binding)
```

(which is what I am suppressing the same way as this is done in a few existing cases)